spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingExtractor.java

```diff
@@ -17,14 +18,23 @@
 	@Override
 	public Span joinTrace(SpanTextMap textMap) {
 		Map<String, String> carrier = TextMapUtil.asMap(textMap);
-		if (!hasHeader(carrier, TraceMessageHeaders.SPAN_ID_NAME)
+		if (Span.SPAN_SAMPLED.equals(carrier.get(TraceMessageHeaders.SPAN_FLAGS_NAME))) {
+			generateIdIfMissing(carrier, TraceMessageHeaders.SPAN_ID_NAME);
```

Comment:
this is fine, though you could make it simpler by using the trace == span id when absent